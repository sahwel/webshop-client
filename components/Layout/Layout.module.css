@tailwind components;


@layer components {
    .layout{
        @apply bg-main-light p-5 min-h-screen text-text-light font-poppins
    }

    :global(.dark) .layout{
        @apply text-text-dark bg-main-dark 
    }
    
}
    
