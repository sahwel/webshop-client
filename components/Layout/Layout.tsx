import { FunctionComponent, ReactNode } from 'react';
import styles from './Layout.module.css';

interface LayoutProps {
  children?: ReactNode;
}

const Layout: FunctionComponent<LayoutProps> = ({ children }) => {
  return <div className={styles.layout}>{children}</div>;
};

export default Layout;
