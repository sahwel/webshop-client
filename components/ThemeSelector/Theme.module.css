
@tailwind components;

@layer components {
    .theme {
       @apply rounded-full p-2  flex justify-center items-center
    }
}