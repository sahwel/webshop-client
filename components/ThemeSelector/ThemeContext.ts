/* eslint-disable no-undef */
import { createContext } from 'react';
import { Themes } from './definitions';

export interface ThemeContextModel {
  setTheme(): void;
  getTheme(): Themes;
}

export const ThemeContext = createContext<ThemeContextModel>(null as any);
