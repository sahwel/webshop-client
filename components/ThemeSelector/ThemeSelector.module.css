
@tailwind components;

@layer components {
    .theme_sececltor {
        @apply grid grid-cols-2 z-50 cursor-pointer bg-white absolute right-5  rounded-full border-2 border-theme-border-light
    }  
    
    .theme_sun{
        @apply text-orange-600 
    }

    .theme_moon{
        @apply text-gray-400 rotate-[255deg]
    }

    .theme_moon_highlited{
        @apply bg-[#303841]
    }

    .theme_sun_highlited{
        @apply bg-orange-200
    }
}