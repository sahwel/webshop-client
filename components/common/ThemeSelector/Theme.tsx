import { FunctionComponent, ReactNode } from 'react';
import cl from 'classnames';
import styles from './theme.module.css';

interface ThemeProps {
  children?: ReactNode;
  isHighlighted: boolean;
  highlightedClass: string;
}

const Theme: FunctionComponent<ThemeProps> = ({
  children,
  isHighlighted,
  highlightedClass,
}) => {
  return (
    <div className={cl({ [highlightedClass]: isHighlighted }, styles.theme)}>
      {children}
    </div>
  );
};

export default Theme;
