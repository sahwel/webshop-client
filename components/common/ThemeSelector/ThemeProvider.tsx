import {
  FunctionComponent,
  ReactNode,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { ThemeContext, ThemeContextModel } from './ThemeContext';
import { Themes } from './definitions';
import Loading from '../Loading/Loading';

interface ThemeProviderProps {
  children?: ReactNode;
}

const ThemeProvider: FunctionComponent<ThemeProviderProps> = ({ children }) => {
  const [currentTheme, setCurrentTheme] = useState<Themes>('light');
  const [isLoaded, setIsLoaded] = useState(false);
  useEffect(() => {
    // todo: get alternative solution
    const theme = localStorage.theme ? localStorage.theme : 'light';

    if (theme === 'dark') {
      document.documentElement.classList.add('dark');
    } else {
      document.documentElement.classList.remove('dark');
    }
    setCurrentTheme(theme);
    setIsLoaded(true);
    return () => {};
  }, []);
  const contextValue: ThemeContextModel = useMemo(
    () => ({
      setTheme() {
        setCurrentTheme((oldTheme) => {
          const newTheme = oldTheme === 'light' ? 'dark' : 'light';

          if (newTheme === 'dark') {
            document.documentElement.classList.add('dark');
          } else {
            document.documentElement.classList.remove('dark');
          }
          localStorage.theme = newTheme;
          return newTheme;
        });
      },
      getTheme() {
        return currentTheme;
      },
    }),
    [currentTheme]
  );
  if (!isLoaded) return null;
  return (
    <ThemeContext.Provider value={contextValue}>
      {children}
    </ThemeContext.Provider>
  );
};

export default ThemeProvider;

export function useTheme() {
  const value = useContext(ThemeContext);
  return value;
}
