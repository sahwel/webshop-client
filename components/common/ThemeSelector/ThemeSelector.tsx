import { SunIcon } from '@heroicons/react/outline';
import { MoonIcon } from '@heroicons/react/solid';
import { useCallback, useLayoutEffect, useState } from 'react';
import cl from 'classnames';
import Theme from './Theme';
import styles from './ThemeSelector.module.css';
import { useTheme } from './ThemeProvider';

const ThemeSelector = () => {
  const siteTheme = useTheme();
  const handleChange = useCallback(() => {
    siteTheme.setTheme();
  }, []);

  return (
    <div className={styles.theme_sececltor} onClick={handleChange}>
      <Theme
        highlightedClass={styles.theme_sun_highlited}
        isHighlighted={siteTheme.getTheme() === 'light'}
      >
        <SunIcon className={cl('icon-6', styles.theme_sun)} />
      </Theme>

      <Theme
        highlightedClass={styles.theme_moon_highlited}
        isHighlighted={siteTheme.getTheme() === 'dark'}
      >
        <MoonIcon
          className={cl('icon-6', styles.theme_moon, {
            ['!text-white ']: siteTheme.getTheme() === 'dark',
          })}
        />
      </Theme>
    </div>
  );
};

export default ThemeSelector;
