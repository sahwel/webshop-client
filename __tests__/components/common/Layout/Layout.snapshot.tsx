import { render, renderHook } from '@testing-library/react';
import Layout from '@/components/Layout/Layout';
import {
  mockUseThemeSelector,
  themeSelectorWrapper,
} from '../../../../testUtils/MockThemeSelector';
import React from 'react';
import { useTheme } from '@/components/common/ThemeSelector/ThemeProvider';

React.useContext = mockUseThemeSelector;

describe('Layout snapshot', () => {
  let _result;
  beforeEach(() => {
    const { result } = renderHook(() => useTheme(), {
      wrapper: themeSelectorWrapper,
    });

    _result = result;
  });
  it('renders LoginFooter unchanged', () => {
    const { container } = render(<Layout children={<p>test children</p>} />);
    expect(container).toMatchSnapshot();
  });
});
