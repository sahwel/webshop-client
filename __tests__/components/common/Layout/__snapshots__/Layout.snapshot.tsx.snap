// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout snapshot renders LoginFooter unchanged 1`] = `
<div>
  <div
    class="layout"
  >
    <p>
      test children
    </p>
  </div>
</div>
`;
